### PR TITLE
feat: add global approval policy and enforcement

### PR DESCRIPTION
**Pull request by eshankvaish**
# Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A
- [ ] Test B

## Checklist

- [ ] The title of the PR states what changed and the related issues number (used for the release note).
- [ ] Does this PR require documentation updates?
- [ ] I've updated documentation as required by this PR.
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/Common/Types.ts [hunk 1]
```diff
@@ -396,12 +363,48 @@ interface ApprovalUserDataType {
     userGroups?: Pick<UserGroupDTO, 'identifier' | 'name'>[]
 }
 
+export interface UserApprovalInfo {
+    requiredCount: number
+    currentCount: number
+    approverList: {
```
Comment (AbhishekA1509) [inline src/Common/Types.ts:369]: Please move to interface

File: src/Common/Types.ts [hunk 2]
```diff
@@ -396,12 +363,48 @@ interface ApprovalUserDataType {
...
+        hasApproved: boolean
+        canApprove: boolean
+        identifier: string
+    }[]
+}
+
+export enum ApprovalConfigDataKindType {
```
Comment (AbhishekA1509) [inline src/Common/Types.ts:376]: Please check with BE if we can extend existing enum of ConfigResourceType which is being used across protection flow

File: src/Common/Types.ts [hunk 3]
```diff
@@ -396,12 +363,48 @@ interface ApprovalUserDataType {
...
+    configSecret = 'CS',
+    deploymentTemplate = 'DEPLOYMENT_TEMPLATE',
+}
+
+export interface ApprovalConfigDataType extends Pick<UserApprovalInfo, 'currentCount' | 'requiredCount'> {
+    kind: ApprovalConfigDataKindType | null
+    anyUserApprovedInfo: UserApprovalInfo
+    specificUsersApprovedInfo: UserApprovalInfo
+    userGroupsApprovedInfo: Pick<UserApprovalInfo, 'currentCount' | 'requiredCount'> & {
+        userGroups: (UserApprovalInfo & {
```
Comment (AbhishekA1509) [inline src/Common/Types.ts:387]: Can move to interface since extending this below

File: src/Common/Common.service.ts [hunk 4]
```diff
@@ -211,110 +189,38 @@ const cdMaterialListModal = ({
...
-            const userGroups = imageApprovalUsersInfo[user] || []
-            userGroups.forEach((group) => {
-                if (!acc[group.identifier]) {
-                    acc[group.identifier] = {}
-                }
-                acc[group.identifier][user] = true
-            })
-            return acc
+        specificUsers: {
+            approverList: sanitizeUserApprovalList(deploymentApprovalInfo?.eligibleApprovers?.anyUsers?.approverList),
```
Comment (AbhishekA1509) [inline src/Common/Common.service.ts:200]: Please move deploymentApprovalInfo?.eligibleApprovers?.anyUsers?.approverList into variable

File: src/Common/Common.service.ts [hunk 5]
```diff
@@ -91,42 +85,27 @@ export function setImageTags(request, pipelineId: number, artifactId: number) {
...
-        ...approvalMetadata,
-        approvedUsersData: approvedUsersData.map((userData) => ({
-            ...userData,
-            userGroups: userData.userGroups?.filter((group) => !!group?.identifier && !!group?.name) ?? [],
-        })),
-        approvalConfig: sanitizeUserApprovalConfig(unsanitizedApprovalConfig),
-    }
-}
+export const sanitizeUserApprovalMetadata = (userApprovalMetadata: UserApprovalMetadataType): UserApprovalMetadataType => ({
+    ...userApprovalMetadata,
```
Comment (AbhishekA1509) [inline src/Common/Common.service.ts:89]: Can de-structure null

File: src/Shared/Components/CICDHistory/service.tsx [hunk 6]
```diff
@@ -42,21 +42,26 @@ const getTriggerDetailsQuery = (fetchIdData) => {
...
-    if (triggerId) {
-        return get(
-            `${ROUTES.APP}/cd-pipeline/workflow/trigger-info/${appId}/${envId}/${pipelineId}/${triggerId}${getTriggerDetailsQuery(fetchIdData)}`,
-        )
-    }
-    return get(
-        `${ROUTES.APP}/cd-pipeline/workflow/trigger-info/${appId}/${envId}/${pipelineId}/last${getTriggerDetailsQuery(fetchIdData)}`,
+    const response = await get<TriggerDetailsResponseType['result']>(
+        triggerId
+            ? `${ROUTES.APP}/cd-pipeline/workflow/trigger-info/${appId}/${envId}/${pipelineId}/${triggerId}${getTriggerDetailsQuery(fetchIdData)}`
```
Comment (AbhishekA1509) [inline src/Shared/Components/CICDHistory/service.tsx:54]: Suggestion: `${ROUTES.APP}/cd-pipeline/workflow/trigger-info/${appId}/${envId}/${pipelineId} is constant and can we create route

File: src/Shared/Components/CICDHistory/TriggerOutput.tsx [hunk 7]
```diff
@@ -703,6 +703,7 @@ const TriggerOutput = ({
             const promotionApprovalMetadata = triggerHistory.get(syncTriggerId)?.promotionApprovalMetadata
             const runSource = triggerHistory.get(syncTriggerId)?.runSource
             const targetConfig = triggerHistory.get(syncTriggerId)?.targetConfig
+            const userApprovalMetadata = triggerHistory.get(syncTriggerId)?.userApprovalMetadata
```
Comment (AbhishekA1509) [inline src/Shared/Components/CICDHistory/TriggerOutput.tsx:706]: Please ensure this is sanitized

File: src/Shared/constants.tsx [hunk 8]
```diff
@@ -504,4 +504,13 @@ export const OPEN_NEW_TICKET = 'https://enterprise.devtron.ai/portal/en/newticke
...
 export const RAISE_ISSUE = 'https://github.com/devtron-labs/devtron/issues/new/choose'
 
+export const TARGET_IDS = {
+    WORKFLOW_EDITOR_CONTAINER: 'workflow-editor-container',
+}
+
+/**
+ * Env ID to get the data for base configuration
+ */
+export const BASE_CONFIGURATION_ENV_ID = -1
```
Comment (AbhishekA1509) [inline src/Shared/constants.tsx:514]: Had made this in DeploymentTemplate in dashboard please replace that as well

**Review by AbhishekA1509 — CHANGES_REQUESTED** [review]

**Review by AbhishekA1509 — DISMISSED** [review]

**Review by AbhishekA1509 — DISMISSED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Review by vivek-devtron — APPROVED** [review]

**Pull request merged**